scrapyd_k8s/log_handler.py

```diff
@@ -0,0 +1,161 @@
+import os
+import threading
+import tempfile
+import logging
+
+from kubernetes import client, watch
+from .object_storage import LibcloudObjectStorage
+
+logger = logging.getLogger(__name__)
+
+watcher_threads = {}
+pod_tmp_mapping = {}
+
+def get_last_n_lines(file_path, num_lines):
+    """Efficiently retrieve the last `num_lines` lines from `file_path`."""
+    try:
+        with open(file_path, 'rb') as f:
+            # Move to the end of the file
+            f.seek(0, os.SEEK_END)
+            file_size = f.tell()
+            block_size = 6144
+            data = b''
+            remaining_lines = num_lines
+            while remaining_lines > 0 and file_size > 0:
+                if file_size - block_size > 0:
+                    f.seek(file_size - block_size)
+                    block_data = f.read(block_size)
+                else:
+                    f.seek(0)
+                    block_data = f.read(file_size)
+                file_size -= block_size
+                data = block_data + data
+                lines_found = data.count(b'\n')
+                # If we've found enough lines, we can stop reading more blocks
+                if lines_found >= remaining_lines:
+                    break
+
+            # Decode the data and split into lines
+            lines = data.decode('utf-8', errors='replace').splitlines()
```

Comment:
Just to be sure: do we know this is `utf-8`?

Reply:
Isn't utf-8 standard? Even windows made utf-8 sort of default, unless, a person uses some old windows machine to host this cluster, then generated files will probably have their famous typewriter coding. Anyway, I would still expect utf-8.

Reply:
Well, yes, it is kind of the default noawadays, but it may vary in certain cases.
It is just a hardcoded assumption that I'd like to check. In production-grade software, I'd expect one to explain why this is utf-8 (e.g. linking to software generating the log, where it is mentioned - I don't think we know that for sure here), or explain why e.g. it doesn't really matter that much (byte-wise comparison would also be ok, we're just interested in newlines and bytes here I think - with e.g. utf-16 this would matter, but for most other regular encodings, this doesn't make a difference).
Or mention that this is an assumption - I think that would be fine here :)